csharp/src/iceboxnet/ServiceManager.cs

```diff
@@ -2,195 +2,125 @@
                 _communicator.GetProperty("Ice.Admin.Endpoints") != null;
 
-            if (_adminEnabled)
+            if (adminEnabled)
             {
-                _adminFacetFilter = new HashSet<string>(_communicator.GetPropertyAsList("Ice.Admin.Facets") ??
-                                                        Array.Empty<string>());
+                _adminFacetFilter = new HashSet<string>(
+                    _communicator.GetPropertyAsList("Ice.Admin.Facets") ?? Array.Empty<string>());
             }
 
             _argv = args;
             _traceServiceObserver = _communicator.GetPropertyAsInt("IceBox.Trace.ServiceObserver") ?? 0;
         }
 
-        public void AddObserver(IServiceObserverPrx? observer, Current current, CancellationToken cancel)
+        public ValueTask AddObserverAsync(IServiceObserverPrx observer, Current current, CancellationToken cancel)
         {
-            // Null observers and duplicate registrations are ignored
-            if (observer != null)
+            lock (_mutex)
             {
-                string[] activeServices;
-                lock (_mutex)
+                if (!_observers.Add(observer))
                 {
-                    try
-                    {
-                        _observers.Add(observer);
-                    }
-                    catch (ArgumentException)
-                    {
-                        return;
-                    }
-
-                    if (_traceServiceObserver >= 1)
-                    {
-                        _logger.Trace("IceBox.ServiceObserver", $"Added service observer {observer}");
-                    }
-
-                    activeServices = _services.Where(info => info.Status == ServiceStatus.Started).Select(
-                        info => info.Name).ToArray();
+                    return default; // ignore duplicates.
                 }
 
-                if (activeServices.Length > 0)
+                if (_traceServiceObserver >= 1)
                 {
-                    _ = StartedAsync(observer, activeServices);
+                    _logger.Trace("IceBox.ServiceObserver", $"added service observer {observer}");
                 }
 
-                async Task StartedAsync(IServiceObserverPrx observer, string[] services)
-                {
-                    try
-                    {
-                        await observer.ServicesStartedAsync(services, cancel: cancel).ConfigureAwait(false);
-                    }
-                    catch (Exception ex)
-                    {
-                        RemoveObserver(observer, ex);
-                    }
-                }
-            }
-        }
+                // Important: IsStarted is called (possibly lazily) with _mutex locked.
 
-        public void Shutdown(Current current, CancellationToken cancel) => _ = _communicator.ShutdownAsync();
+                IEnumerable<string> startedServices = _services.Where(info => info.IsStarted).Select(info => info.Name);
```

Comment:
The enumerable is evaluated while we hold the mutex. The latest time it can get evaluated is when we marshal the sequence, which occurs while we hold the mutex. See `StartedAsync`. I'll add a comment.